### ISSUE TITLE
How to pass strict-ssl to Yarn/NPM?

### ISSUE THREAD

**Issue by canthony** [devsoap/gradle-vaadin-flow#186]
**TLDR**: How can I configure npm/yarn to use strict-ssl false at the **project** level? And, might it not be worth adding this to the wiki at some point, to save others the pain?!

**Desktop (please complete the following information):**
 - OS: Windows
 - JDK version: 8.0
 - Gradle version: 5.0
 - Vaadin Plugin version:  1.0.0.M6

**Describe the bug**
We're working behind a corporate proxy. It does packet inspections ->the SSL Certificate is modified. Yuck, yes :  but we can't change that. We need to tell everything to ignore certificate mismatch errors.

**To Reproduce**
1. Follow the instructions in the wiki to create the project
2. Run ./gradlew clean build
3. This produces the following
~~~
...
> Configure project :
Using Gradle Vaadin Flow Plugin 1.0.0-RC6
vaadin.version is not set, falling back to latest Vaadin version
Allow Vaadin to gather usage statistics by setting vaadin.submitStatistics=true (hide this message by setting it to false)

> Task :clean
> Task :compileJava
> Task :vaadinConvertStyleCssToHtml
> Task :nodeSetup

> Task :yarnSetup FAILED
npm ERR! code SELF_SIGNED_CERT_IN_CHAIN
npm ERR! errno SELF_SIGNED_CERT_IN_CHAIN
npm ERR! request to https://registry.npmjs.org/yarn failed, reason: self signed certificate in certificate chain

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Charles\AppData\Roaming\npm-cache\_logs\2019-01-21T14_27_54_145Z-debug.log
~~~

**Expected behavior**
This rather intimidating to the uninitiated! I don't know anything about Vaadin Flow, Node, Bower, Yarn or NPM - I just wanted to get started with a simple project...
After a bit of googling, I found that if I create a file in ~/.npmrc with "strict-ssl=false"
~~~
cat > ~/.npmrc <<EOF
strict-ssl=false
EOF
~~~ 
and Rerun, everything works just fine.
~~~
charles@PC002:~/exampleapp$ ./gradlew clean build
...
Using Gradle Vaadin Flow Plugin 1.0.0-RC6
vaadin.version is not set, falling back to latest Vaadin version
Allow Vaadin to gather usage statistics by setting vaadin.submitStatistics=true (hide this message by setting it to false)

> Task :clean
> Task :compileJava
> Task :vaadinConvertStyleCssToHtml
> Task :nodeSetup UP-TO-DATE

> Task :yarnSetup
C:\Users\Charles\exampleapp\.gradle\yarn\yarn-v1.12.3\yarn -> C:\Users\Charles\exampleapp\.gradle\yarn\yarn-v1.12.3\node_modules\yarn\bin\yarn.js
C:\Users\Charles\exampleapp\.gradle\yarn\yarn-v1.12.3\yarnpkg -> C:\Users\Charles\exampleapp\.gradle\yarn\yarn-v1.12.3\node_modules\yarn\bin\yarn.js
+ yarn@1.12.3
added 1 package in 0.752s

> Task :vaadinInstallYarnDependencies SKIPPED
> Task :vaadinInstallBowerDependencies SKIPPED
> Task :vaadinConvertGroovyTemplatesToHtml SKIPPED
> Task :vaadinTranspileDependencies SKIPPED
> Task :vaadinAssembleClient SKIPPED
> Task :vaadinPluginVersionCheck
> Task :processResources NO-SOURCE
> Task :classes
> Task :jar
> Task :assemble
> Task :compileTestJava NO-SOURCE
> Task :processTestResources NO-SOURCE
> Task :testClasses UP-TO-DATE
> Task :test NO-SOURCE
> Task :check UP-TO-DATE
> Task :build
~~~

However : I don't want to have to create a ~/.npmrc file on every single machine that does a build - that'd be every developer machine, and every jenkins/bamboo build agent. 

Ideally I'd just be able to add "strict-ssl false" to the project level .npmrc or .yarnrc - but neither of those things seems to work (I started to look at bug reports for both projects, but my eyes glazed over)

TLDR; how can I configure npm/yarn to have strict-ssl false at the project level? And, might it not be worth adding this to the wiki at some point, to save others the pain!

**Comment by johndevs** [comment]
As a workaround you could try adding this to your **build.gradle**:
```
vaadinInstallYarnDependencies {
    afterEvaluate {
        yarnRunner.generateYarnRc()
        yarnrc << 'strict-ssl false'
    }
}
```
It is rather ugly, but it should do the job for now on project level so you don't have to manually configure any machines (yuck!).

I'll improve this a bit for the next version so customizing Yarn would be easier in the future. 

PS: Welcome to the dark side of frontend tooling ;)

**Comment by canthony** [comment]
@johndevs Thanks John; sorry to say, it hasn't had the right effect. It's entirely possible it's a Yarn issue, I guess. 

FWIW, you should be able to reproduce by sticking a dev proxy in the way, like Fiddler.

**build.gradle:**
```groovy
plugins {
  'org.gretty' version '2.3.1'
  id 'com.devsoap.vaadin-flow' version '1.0.0.RC6'
}

vaadin {
  version = '10.0.10'
  submitStatistics = false
  autoconfigure()
}

vaadinInstallYarnDependencies {
  afterEvaluate {
    yarnRunner.generateYarnRc()
    yarnrc << 'strict-ssl false'
  }
}
```
**output**
```
charles@PC002:~/exampleapp$ ./gradlew build

> Configure project :
Using Gradle Vaadin Flow Plugin 1.0.0-RC6

> Task :compileJava UP-TO-DATE
> Task :vaadinConvertStyleCssToHtml UP-TO-DATE
> Task :nodeSetup UP-TO-DATE

> Task :yarnSetup FAILED
npm ERR! code SELF_SIGNED_CERT_IN_CHAIN
npm ERR! errno SELF_SIGNED_CERT_IN_CHAIN
npm ERR! request to https://registry.npmjs.org/yarn failed, reason: self signed certificate in certificate chain

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Charles\AppData\Roaming\npm-cache\_logs\2019-01-22T07_54_49_608Z-debug.log

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':yarnSetup'.
> Process 'command 'C:\Users\Charles\exampleapp\.gradle\nodejs\node-v10.14.1-win-x64\node.exe'' finished with non-zero exit value 1

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 3s
4 actionable tasks: 1 executed, 3 up-to-date
```

**Comment by johndevs** [comment]
hmm, according to https://github.com/yarnpkg/yarn/issues/980#issuecomment-257413314 it should work. 

Just to verify, did you do a ``gradle clean`` before running the above so ``.yarnrc`` was re-created? You could also take a peek into ``build/frontend`` and see how the ``.yarnrc`` file looks like and verify you go the parameter there?

**Comment by johndevs** [comment]
I noticed a similar thing while fixing it. Too bad that workaround didn't work as intended :(

But I made a proper fix for it in #186. 

It will allow you configure any yarnrc property via
```
vaadinClientDependencies {
     customYarnProperties = ['strict-ssl' : false ]
}
```
This will then be used when the file is generated so it shouldn't have the problems the workaround has. 

I'll include it in the next RC release.

**Comment by johndevs** [comment]
RC7 is now released with that property, could you try it out and see if it works for you?

**Comment by canthony** [comment]
Hi John, 

Sorry for not responding earlier - I was away.

I've just tried with RC7, and I'm afraid it's still not quite there.

```
> Configure project :
Using Gradle Vaadin Flow Plugin 1.0.0-RC7

> Task :compileJava UP-TO-DATE
> Task :vaadinConvertStyleCssToHtml UP-TO-DATE
> Task :nodeSetup UP-TO-DATE

> Task :yarnSetup FAILED
npm ERR! code SELF_SIGNED_CERT_IN_CHAIN
npm ERR! errno SELF_SIGNED_CERT_IN_CHAIN
npm ERR! request to https://registry.npmjs.org/yarn failed, reason: self signed certificate in certificate chain

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Charles\AppData\Roaming\npm-cache\_logs\2019-02-04T08_13_31_037Z-debug.log

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':yarnSetup'.
> Process 'command 'C:\Users\Charles\exampleapp\.gradle\nodejs\node-v10.14.1-win-x64\node.exe'' finished with non-zero exit value 1
```

However, with my limited (virtually non-existent) understanding of what's going on, I think I can see the issue.

We've been generating/modifying .yarnc - and yet, If I look at the debug log mentioned above, I think the problem here is being generated when we (well, the gradle node plugin) attempts to *install* yarn. 
(The log, included below, is basically running "node.exe npm-cli.js install yarn@1.13.3")

In other words, your changes are telling *yarn* to skip SSL checks, but we probably need to tell *node* this (too?)

Cheers,

Charles.

```
harles@PC002:~/exampleapp$ cat /cygdrive/c/Users/Charles/AppData/Roaming/npm-cache/_logs/2019-02-04T08_13_31_037Z-debug.log
0 info it worked if it ends with ok
1 verbose cli [ 'C:\\Users\\Charles\\exampleapp\\.gradle\\nodejs\\node-v10.14.1-win-x64\\node.exe',
1 verbose cli   'C:\\Users\\Charles\\exampleapp\\.gradle\\nodejs\\node-v10.14.1-win-x64\\node_modules\\npm\\bin\\npm-cli.js',
1 verbose cli   'install',
1 verbose cli   '--global',
1 verbose cli   '--no-save',
1 verbose cli   '--prefix',
1 verbose cli   'C:\\Users\\Charles\\exampleapp\\.gradle\\yarn\\yarn-v1.12.3',
1 verbose cli   'yarn@1.12.3' ]
2 info using npm@6.4.1
3 info using node@v10.14.1
4 verbose npm-session 0ba19830d85c611b
5 silly install loadCurrentTree
6 silly install readGlobalPackageData
7 silly fetchPackageMetaData error for yarn@1.12.3 request to https://registry.npmjs.org/yarn failed, reason: self signed certificate in certificate chain
8 timing stage:rollbackFailedOptional Completed in 1ms
9 timing stage:runTopLevelLifecycles Completed in 456ms
10 verbose type system
11 verbose stack FetchError: request to https://registry.npmjs.org/yarn failed, reason: self signed certificate in certificate chain
11 verbose stack     at ClientRequest.req.on.err (C:\Users\Charles\exampleapp\.gradle\nodejs\node-v10.14.1-win-x64\node_modules\npm\node_modules\node-fetch-npm\src\index.js:68:14)
11 verbose stack     at ClientRequest.emit (events.js:182:13)
11 verbose stack     at TLSSocket.socketErrorListener (_http_client.js:391:9)
11 verbose stack     at TLSSocket.emit (events.js:182:13)
11 verbose stack     at emitErrorNT (internal/streams/destroy.js:82:8)
11 verbose stack     at emitErrorAndCloseNT (internal/streams/destroy.js:50:3)
11 verbose stack     at process._tickCallback (internal/process/next_tick.js:63:19)
12 verbose cwd C:\Users\Charles\exampleapp
13 verbose Windows_NT 10.0.17134
14 verbose argv "C:\\Users\\Charles\\exampleapp\\.gradle\\nodejs\\node-v10.14.1-win-x64\\node.exe" "C:\\Users\\Charles\\exampleapp\\.gradle\\nodejs\\node-v10.14.1-win-x64\\node_modules\\npm\\bin\\npm-cli.js" "install" "--global" "--no-save" "--prefix" "C:\\Users\\Charles\\exampleapp\\.gradle\\yarn\\yarn-v1.12.3" "yarn@1.12.3"
15 verbose node v10.14.1
16 verbose npm  v6.4.1
17 error code SELF_SIGNED_CERT_IN_CHAIN
18 error errno SELF_SIGNED_CERT_IN_CHAIN
19 error request to https://registry.npmjs.org/yarn failed, reason: self signed certificate in certificate chain
20 verbose exit [ 1, true ]
```

**Comment by johndevs** [comment]
From https://github.com/npm/npm/issues/7519:
> Sometimes Windows users have an SSL-intercepting proxy; npm detects this and complains. You may need to set https_proxy specially, depending on your local network environment. See https://github.com/npm/npm/wiki/Troubleshooting#ssl-error for an extended troubleshooting guide to common SSL-related errors.

According to that you might be missing the https proxy configuration. You will need both a http and a https proxy configuration, theres instructions here how to define them https://docs.gradle.org/current/userguide/build_environment.html#sec:accessing_the_web_via_a_proxy.

**Comment by canthony** [comment]
The Gradle https proxy configuration is good, and working – that’s why it’s complaining about the "self signed certificate in certificate chain" (the irritating corporate proxy is doing SSL Packet Inspections, and thus replacing the destination site’s SSL Certificate with its own. It's not actually self-signed, but that's not the problem - it's that it doesn't match).

We need to tell node to disable ssl checking, in much the same way that you have with yarn. From a comment on [npm/npm#7519 ](https://github.com/npm/npm/issues/7519#issuecomment-390561629): 

> npm config set strict-ssl false

I guess that there is/could be  a project level npm config file that we could generate/add the "strict-ssl false" property to.

**Comment by johndevs** [comment]
According to https://docs.npmjs.com/files/npmrc.html you should be able to create a project-specific .npmrc file.

You can try this out and see if it works, add a file called **.npmrc** under **\<project dir\>/build/frontend/** with the following content:
```
strict-ssl=false
proxy=<http proxy>
https-proxy=<https proxy>
cafile=<path/to/certificate/pem>
```
Not sure you will need the cafile or not. If it works I'll automate it so that the plugin automatically generates the file in the future.

**Comment by johndevs** [comment]
Okay, I think you're right, its the npm install of yarn that is failing, not yarn itself.

I was able to configure npm via the following workaround by adding the following magic to my build.gradle:

```
project.afterEvaluate {
    def rc = Paths.get(project.buildDir.canonicalPath, 'frontend', '.npmrc').toFile()
    if(rc.exists()) rc.delete()

    // Add your settings here:
    rc << "proxy=http://localhost:4873\n"
    rc << "https-proxy=http://localhost:4873\n"
    rc << "strict-ssl=false\n"

    yarnSetup.args -= '--global'
    yarnSetup.args.add(2, '--userconfig')
    yarnSetup.args.add(3, rc.canonicalPath)
}
```
This of course is very ugly, I'll fix this in the plugin so it works much nicer in the same way as the yarnrc configuration is done but I hope this will get you forward again :)

**Comment by johndevs** [comment]
Right, finally got that PR to pass the tests and merged :grin: 

So in RC8, you will be able to configure NPM in the same way as Yarn by adding the following into build.gradle:
```
vaadinClientDependencies {
    customYarnProperties = ['strict-ssl', false]
    customNpmProperties = ['strict-ssl', false]
}
```

**Comment by canthony** [comment]
Thanks John.

I'll give it a spin in RC8 when you release it ; I know it's a PITA affecting only a small number of people, but I can also see that simple properties configuration  might be useful for other fiddly bits and pieces later on, too.

Cheers,

Charles.

**Comment by canthony** [comment]
Hi John - A Week later, I've tested this, and it's all good. Thanks!
